### PR TITLE
Truss branch - Deal with test fail all over the place

### PR DESCRIPTION
**Pull request by briehl**
# Description of PR purpose/changes

Yet again, the front end tests have been flaky and failing consistently, causing much wailing and gnashing of teeth in the Data Upload team.
This time, I think I found one of the more insidious root causes.
When the Runtime component gets made, it does two things.
1. Puts a copy of the root bus into `window.kbaseRuntime`.
2. Creates a new `Clock` object.

To deal with emptying out the master bus so it doesn't pollute tests that expect a clean, unlinked bus, `window.kbaseRuntime` was commonly set to `null` between tests. However, when the `Clock` is created, it sets up an interval that emits a `clock-tick` bus message every second on that bus. Clearing out `window.kbaseRuntime` doesn't remove the interval, or the invisible handle that the `Clock` keeps on the bus, so it wasn't really all that effective. In fact, this resulted in the creation of a new `Bus` and a new `Clock` (and a new interval) for most every test.

I added a little debug code, and at one point I counted 2680 clocks running simultaneously, all spamming bus messages on various intervals. I hypothesized that this was one of the problems causing the strange failures, many of which had to do with bus messages either being lost or misinterpreted.

I refactored the Runtime module to put the generated Clock into `window.kbaseRuntime`, and added additional code to destroy the Runtime bus element on request. Finally, I added a `clearRuntime()` function to `test/unit/testUtil.js` that removes all of the bits of `window.kbaseRuntime` if present. This now gets run as either an `afterEach` or `afterAll` for most (if not all) tests. It's unclear when the module being tested has some dependency that invokes `Runtime.make()` so this was added all over the place to be safe. It's probably overkill.

During this process, I also identified a few other issues that were causing problems. They're mentioned in some PR comments, but here's some highlights
* Disabled many of the root bulkImportCell tests for now, as they're failing because of how the Promise structure is made (or, rather, not made) when the BulkImportCell is created. That'll be a later PR.
* Some StagingAreaViewer DataTables functionality was infesting the global `$.fn.DataTable` object and bleeding over into other tests. Also affected `kbaseNarrativeStatingDataTab.js`
* At least one test was depending on a race condition. I'm thinking this previously passed more often due to the slowdown from zillions of buses and intervals running at once.
* Some of the Runtime object tests were failing when run in a different order - global state was once again being modified by another test suite (`upaAPI`) without putting it back in order.

In the end, I'm hopeful this will make the front end unit tests pass more consistently when expected. Or at least fail due to structural errors less often.

Side note - the Runtime stuff above doesn't affect actual running in the page as there aren't any modules that destroy the Runtime. Thus, during normal page activity, there's only one Runtime sitting in global space and not being removed. This is really only an issue of our current test apparatus that doesn't guarantee a cleared environment for each suite.

# Jira Ticket / Issue #
e.g. https://kbase-jira.atlassian.net/browse/DATAUP-X
- [n/a] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
- [x] Tests pass locally and in GitHub Actions
- [ ] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [n/a] Any dependent changes have been merged and published in downstream modules
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook
- [n/a] (Python) I have run Black and Flake8 on changed Python code manually or with a git precommit hook

### CODE REVIEW

File: kbase-extension/static/kbase/js/common/monoBus.js [hunk 1]
```diff
@@ -42,6 +42,15 @@ define(['uuid', 'bluebird', './lang', './unodep'], (Uuid, Promise, lang, utils)
             }
         }
 
+        /**
+         * Clears all variables and the timer, if it exists
+         */
+        function destroy() {
+            if (timer) {
+                clearTimeout(timer);
```
Comment (briehl) [inline kbase-extension/static/kbase/js/common/monoBus.js:50]: There's a "timer" that just waits for the current tick to finish. It's really a `setTimeout(()=>{}, 0)` function. I figured this would be prudent, just in case.

I guess it could be an edge case where that timeout fires right at the end of a test, then sends messages off to things that have been reaped. I'm not sure how to test for that, though.

File: kbase-extension/static/kbase/js/common/runtime.js [hunk 2]
```diff
@@ -30,15 +30,22 @@ define([
...
                 bus: theBus,
                 env: Props.make({}),
+                clock,
             };
         }
 
         function destroy() {
-            clock.stop();
+            if (clock) {
+                clock.stop();
```
Comment (briehl) [inline kbase-extension/static/kbase/js/common/runtime.js:39]: actually stop the clock tick timer!

File: kbase-extension/static/kbase/js/widgets/function_output/kbaseAttributeMapping.js [hunk 3]
```diff
@@ -14,11 +15,10 @@ define([
     'datatables.net-buttons-print',
     'knhx',
     'widgetMaxWidthCorrection',
-], (KBWidget, $, Config, kbaseAuthenticatedWidget) => {
+], (KBWidget, $, Config, Workspace, Runtime) => {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/function_output/kbaseAttributeMapping.js:18]: The tests around this suite were consistently failing. In digging through why, I found a few oddities that were worth addressing.

Mainly, it was a race condition for starting and running the tests, due to the use of `kbaseAuthenticatedWidget` as a parent here.  Also, because of the call to `renderAndLoad` during `init` and in the `loggedInCallback`, that was being called twice.

File: kbase-extension/static/kbase/js/widgets/function_output/kbaseAttributeMapping.js [hunk 4]
```diff
@@ -5,7 +5,8 @@ define([
     'kbwidget',
     'jquery',
     'narrativeConfig',
-    'kbaseAuthenticatedWidget',
+    'kb_service/client/workspace',
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/function_output/kbaseAttributeMapping.js:8]: Once other things were fixed, this started failing due to an expectation that there would be a globally available Workspace client. Now it uses the Promisified one from `kb_service`

File: kbase-extension/static/kbase/js/widgets/function_output/kbaseAttributeMapping.js [hunk 5]
```diff
@@ -122,26 +115,32 @@ define([
                 columns: cols,
                 data: rows,
             };
-            const ConditionsTable = $tbl.DataTable(tblSettings);
-            //ConditionsTable.draw();
+            this.conditionsTable = $tbl.DataTable(tblSettings);
         },
 
         renderError: function (error) {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/function_output/kbaseAttributeMapping.js:121]: This whole function could probably be supplanted with the ErrorDisplay module, but I'm already way out of scope here. Future work.

File: kbase-extension/static/kbase/js/widgets/function_output/kbaseAttributeMapping.js [hunk 6]
```diff
@@ -152,16 +151,13 @@ define([
...
             this.$messagePane.empty();
         },
 
-        loggedInCallback: function (event, auth) {
-            if (this.token == null) {
-                this.token = auth.token;
-                this.renderAndLoad();
+        destroy: function () {
+            if (this.conditionsTable) {
+                this.conditionsTable.destroy();
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/function_output/kbaseAttributeMapping.js:159]: I put this in while investigating other Datatables things.

File: kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeStagingDataTab.js [hunk 7]
```diff
@@ -52,11 +52,21 @@ define([
...
+                console.error('Call "render" before "activate"');
+            }
         },
 
         deactivate: function () {
-            this.stagingAreaViewer.deactivate();
+            if (this.stagingAreaViewer) {
+                this.stagingAreaViewer.deactivate();
+            }
+            if (this.updateTimeout) {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeStagingDataTab.js:66]: The staging data tab could probably use some restructuring. But this was being problematic.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 8]
```diff
@@ -97,6 +97,10 @@ define([
                 clearInterval(this.refreshInterval);
                 this.refreshInterval = undefined;
             }
+            if (this.fullDataTable) {
```
Comment (briehl) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:100]: We should clear the table out of memory (global? It's hard to say with jquery).
Also, the search extension function (used to filter down to the current directory) needs to be `pop`ped.

File: test/unit/spec/api/authSpec.js [hunk 9]
```diff
@@ -140,6 +140,8 @@ define(['api/auth', 'narrativeConfig', 'uuid'], (Auth, Config, Uuid) => {
     ];
 
     describe('Test the Auth API module', () => {
+        afterAll(() => TestUtil.clearRuntime());
```
Comment (briehl) [inline test/unit/spec/api/authSpec.js:143]: This is an idiom that'll show up throughout this PR.

Either adding `TestUtil.clearRuntime` to `afterAll` or `afterEach`, whichever seemed most appropriate.

File: test/unit/spec/api/upaApiSpec.js [hunk 10]
```diff
@@ -157,6 +162,7 @@ define(['api/upa', 'narrativeConfig'], (UpaApi, Config) => {
         });
 
         it('Should fail if the workspace id cannot be found.', () => {
+            const prevWsId = Config.get('workspaceId');
```
Comment (briehl) [inline test/unit/spec/api/upaApiSpec.js:165]: Not storing and resetting the globally known `workspaceId` was occasionally making failures. This really shouldn't muck with the global space anyway, but this is a fix.

File: test/unit/spec/appWidgets/input/intInputSpec.js [hunk 11]
```diff
@@ -35,7 +39,8 @@ define(['widgets/appWidgets2/input/intInput', 'common/runtime'], (IntInput, Runt
 
         afterEach(() => {
             bus.stop();
-            window.kbaseRuntime = null;
+            runtime.destroy();
```
Comment (briehl) [inline test/unit/spec/appWidgets/input/intInputSpec.js:42]: Any tests that declare `runtime` in the suite also gets to keep the `runtime` so it can be `destroy`ed directly.

File: test/unit/spec/appWidgets/view/dynamicDropdownViewSpec.js [hunk 12]
```diff
@@ -8,83 +8,85 @@ define(['widgets/appWidgets2/view/dynamicDropdownView', 'common/runtime', 'testU
...
-        this.node.remove();
-        window.kbaseRuntime = null;
-    });
-
-    it('should have a valid constructor', function () {
-        const widget = DynamicDropdownView.make({
-            parameterSpec: this.paramSpec,
-            bus: this.bus,
-            initialValue: DEFAULT_VALUE,
+    describe('Dynamic Dropdown View tests', () => {
```
Comment (briehl) [inline test/unit/spec/appWidgets/view/dynamicDropdownViewSpec.js:11]: These weren't inside a `describe` block, so added one here. The rest is unchanged, save for some `clearRuntime` calls.

File: test/unit/spec/appWidgets/view/selectViewSpec.js [hunk 13]
```diff
@@ -41,107 +41,111 @@ define(['widgets/appWidgets2/view/selectView', 'common/runtime', 'testUtil'], (
...
-
-    it('should have a valid constructor', function () {
-        const widget = SelectView.make({
-            parameterSpec: this.paramSpec,
-            bus: this.bus,
-            initialValue: DEFAULT_VALUE,
-        });
-        ['start', 'stop'].forEach((fn) => {
-            expect(widget[fn]).toEqual(jasmine.any(Function));
+    describe('Select view tests', () => {
```
Comment (briehl) [inline test/unit/spec/appWidgets/view/selectViewSpec.js:44]: Also wasn't in a `describe` block. And added `clearRuntime`.

File: test/unit/spec/common/monoBus-Spec.js [hunk 14]
```diff
@@ -415,194 +419,156 @@ define(['common/monoBus'], (Bus) => {
...
     });
 
-    it('Send and receive a test based message over a named channel using a connection', (done) => {
-        const bus = Bus.make(),
-            connection = bus.connect();
-
-        connection.channel('test').on('my-message', (message) => {
-            expect(message.msg).toEqual('greetings');
-            done();
+    describe('Bus functions', () => {
```
Comment (briehl) [inline test/unit/spec/common/monoBus-Spec.js:422]: There were some floating functions here, I put them in a `describe` block.

File: test/unit/spec/common/validate2-Spec.js [hunk 15]
```diff
@@ -322,227 +327,6 @@ define([
             });
         }
 
-        // function runTestsx(method, tests) {
```
Comment (briehl) [inline test/unit/spec/common/validate2-Spec.js:325]: So much commented code!

File: test/unit/spec/function_output/kbaseAttributeMapping-spec.js [hunk 16]
```diff
@@ -1,30 +1,33 @@
-define(['jquery', 'kbaseAttributeMapping', 'base/js/namespace', 'narrativeMocks'], (
-    $,
-    kbaseAttributeMapping,
-    Jupyter,
-    Mocks
-) => {
+define([
```
Comment (briehl) [inline test/unit/spec/function_output/kbaseAttributeMapping-spec.js:1]: See also the `kbaseAttributeMapping.js` code.
Refactored these to work consistently with the other changes.

I also tested them against some sample data. They're KBaseExperiments.AttributeMapping objects.

File: test/unit/spec/function_output/modeling/msPathway-spec.js [hunk 17]
```diff
@@ -1,23 +1,11 @@
...
     'use strict';
+
     describe('Test the msPathway widget', () => {
+        afterAll(() => TestUtil.clearRuntime());
+
         it('Should load the module', () => {
             expect(ModelSeedPathway).toEqual(jasmine.any(Function));
         });
-
-        xit('Should do something or other', () => {
```
Comment (briehl) [inline test/unit/spec/function_output/modeling/msPathway-spec.js:8]: Doesn't seem to do anything, and has been x'd out forever, so gone.

File: test/unit/spec/narrative_core/kbaseNarrativeStagingDataTab-spec.js [hunk 18]
```diff
@@ -47,9 +52,23 @@ define(['jquery', 'kbaseNarrativeStagingDataTab', 'base/js/namespace'], (
...
                         size: 49233,
                         source: 'KBase upload',
+                        isFolder: false,
                     },
                 ]),
             });
+            const mappings = {
+                mappings: [null, null],
+            };
+            jasmine.Ajax.stubRequest(
```
Comment (briehl) [inline test/unit/spec/narrative_core/kbaseNarrativeStagingDataTab-spec.js:62]: This wasn't being mocked, so was actually trying to call out, which was a big problem.

File: test/unit/spec/narrative_core/kbaseNarrativeStagingDataTab-spec.js [hunk 19]
```diff
@@ -138,13 +159,12 @@ define(['jquery', 'kbaseNarrativeStagingDataTab', 'base/js/namespace'], (
...
-            spyOn(stagingWidget.stagingAreaViewer, 'render');
-            for (let i = 0; i < 100; i++) {
+            spyOn(stagingWidget.stagingAreaViewer, 'updateView');
+            for (let i = 0; i < 50; i++) {
                 stagingWidget.uploadWidget.dropzone.emit('complete', { name: 'foo', size: 12345 });
             }
             jasmine.clock().tick(stagingWidget.minRefreshTime + 100);
-            expect(stagingWidget.stagingAreaViewer.render.calls.count()).toEqual(1);
-
+            expect(stagingWidget.stagingAreaViewer.updateView.calls.count()).toEqual(2);
```
Comment (briehl) [inline test/unit/spec/narrative_core/kbaseNarrativeStagingDataTab-spec.js:167]: After some changes, this calls once on the first request as a result of the dropzone event, then once more after the forced clock tick.

File: test/unit/spec/nbextensions/bulkImportCell/bulkImportCell-spec.js [hunk 20]
```diff
@@ -17,25 +18,30 @@ define([
         fakeSpecs = {
             someApp:
                 TestAppObj.app.specs['kb_uploadmethods/import_fastq_sra_as_reads_from_staging'],
-        },
-        runtime = Runtime.make();
+        };
 
-    describe('test the bulk import cell module', () => {
+    xdescribe('test the bulk import cell module', () => {
```
Comment (briehl) [inline test/unit/spec/nbextensions/bulkImportCell/bulkImportCell-spec.js:23]: Skipping these for now.

Other tests happen to cover ~1/2 of the module, though.

File: test/unit/spec/nbextensions/bulkImportCell/fileTypePanel-spec.js [hunk 21]
```diff
@@ -1,153 +0,0 @@
-define([
```
Comment (briehl) [inline test/unit/spec/nbextensions/bulkImportCell/fileTypePanel-spec.js:1]: This is a duplicate of the tests in `test/unit/spec/nbextensions/bulkImportCell/tabs/fileTypePanel-spec.js` (it was moved, but this file wasn't removed. Now it is!)

File: test/unit/spec/userMenuSpec.js [hunk 22]
```diff
@@ -122,20 +123,18 @@ define([
                         // expect to see the modal appear now
                         expect(document.querySelector(modalQuerySelector)).not.toBeNull();
 
-                        // click the appropriate button for this test
```
Comment (briehl) [inline test/unit/spec/userMenuSpec.js:125]: This was actually a race condition that started failing after fixing other things.
The check for the style element on line 131 was failing, as it was already removed! That's the final thing that *should* happen, so this test now skips that intermediate part.

**Comment by briehl** [comment]
All checks pass! I missed all those green checkmarks. 😬

File: kbase-extension/static/kbase/js/common/runtime.js [hunk 23]
```diff
@@ -30,15 +30,22 @@ define([
...
+            }
             $(document).off('dataUpdated.Narrative');
+            if (theBus) {
+                theBus.destroy();
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/runtime.js:43]: 🎉

File: test/unit/spec/api/authSpec.js (hunk 9 above)
Comment (ialarmedalien) [reply test/unit/spec/api/authSpec.js:143]: I wish it was possible to have an `afterAll` for every test file. 😖

File: test/unit/spec/appWidgets/input/intInputSpec.js (hunk 11 above)
Comment (ialarmedalien) [reply test/unit/spec/appWidgets/input/intInputSpec.js:42]: was just about to ask about that!

File: test/unit/karma.conf.js [hunk 24]
```diff
@@ -132,6 +132,6 @@ module.exports = function (config) {
             '/static/kbase/config': '/base/kbase-extension/static/kbase/config',
             '/test/': '/base/test/',
         },
-        concurrency: Infinity,
+        concurrency: 1,
```
Comment (ialarmedalien) [inline test/unit/karma.conf.js:135]: is this left in deliberately?

Comment (briehl) [reply test/unit/karma.conf.js:135]: Oops. Nope. Thanks for catching.

File: test/unit/spec/common/appState-Spec.js [hunk 25]
```diff
@@ -240,6 +241,9 @@ define(['common/fsm'], (Fsm) => {
             }
             expect(alive).toBeTruthy();
         });
+        afterEach(() => {
+            TestUtil.clearRuntime();
+        });
```
Comment (ialarmedalien) [inline test/unit/spec/common/appState-Spec.js:246]: hmmm, who could be responsible for the comment cruft below these lines?!

Comment (briehl) [reply test/unit/spec/common/appState-Spec.js:246]: I WONDER.

File: test/unit/spec/common/runtime-spec.js [hunk 26]
```diff
@@ -21,19 +26,23 @@ define(['common/runtime', 'base/js/namespace', 'narrativeConfig'], (Runtime, Jup
...
         });
 
         beforeEach(function () {
             this.runtime = Runtime.make();
         });
 
-        afterEach(() => {
-            window.kbaseRuntime = null;
+        afterEach(function () {
+            this.runtime.destroy();
```
Comment (ialarmedalien) [inline test/unit/spec/common/runtime-spec.js:43]: I believe the `this` context has already been destroyed by the time you get to the `afterEach` function. I don't know if you want to be extra safe and assign a `runtime` variable so you can explicitly run `runtime.destroy()`?

File: test/unit/spec/common/validate2-Spec.js (hunk 15 above)
Comment (ialarmedalien) [reply test/unit/spec/common/validate2-Spec.js:325]: 😖

File: test/unit/spec/function_output/kbaseAttributeMapping-spec.js [hunk 27]
```diff
@@ -1,30 +1,33 @@
...
         afterEach(() => {
             Mocks.clearAuthToken();
             Jupyter.narrative = null;
             jasmine.Ajax.uninstall();
-            $div.remove();
+            container.remove();
+            TestUtil.clearRuntime();
         });
 
-        it('Should properly render AttributeMapping data', (done) => {
```
Comment (ialarmedalien) [inline test/unit/spec/function_output/kbaseAttributeMapping-spec.js:27]: I thought we had got rid of all those `done`s! Sneaky little so-and-sos...

File: test/unit/spec/function_output/kbaseAttributeMapping-spec.js [hunk 28]
```diff
@@ -58,7 +61,7 @@ define(['jquery', 'kbaseAttributeMapping', 'base/js/namespace', 'narrativeMocks'
                 ],
             };
             // this code is more of less ignored, because two WS calls are made both can't be stubbed
```
Comment (ialarmedalien) [inline test/unit/spec/function_output/kbaseAttributeMapping-spec.js:63]: I remember this bit of code -- if you run it without an internet connection you get workspace connection failures. 😖

File: test/unit/spec/function_output/modeling/msPathway-spec.js (hunk 17 above)
Comment (ialarmedalien) [reply test/unit/spec/function_output/modeling/msPathway-spec.js:8]: I tried to fix it at some point but it seemed irreparably broken. Best deleted!

File: test/unit/spec/narrative_core/kbaseNarrativeStagingDataTab-spec.js (hunk 18 above)
Comment (ialarmedalien) [reply test/unit/spec/narrative_core/kbaseNarrativeStagingDataTab-spec.js:62]: 😖

File: test/unit/spec/appWidgets/input/intInputSpec.js (hunk 11 above)
Comment (briehl) [reply test/unit/spec/appWidgets/input/intInputSpec.js:42]: I wasn't super specific about it. There's a number of places that do `Runtime.make().config('blahblah')` inline in a test. That *could* be declared as a runtime object then destroyed properly, but that would've made an even larger PR...

File: test/unit/spec/common/runtime-spec.js (hunk 26 above)
Comment (briehl) [reply test/unit/spec/common/runtime-spec.js:43]: Yeah, I'll do that. Didn't realize `this` is gone by the `afterEach` function. That's a little confusing.

**Comment by ialarmedalien** [comment]
Oh no! Just when the tests were looking so stable...

**Comment by briehl** [comment]
Yeah, I see it. I think it's the bulk import configure tab doing strange things. Or, rather, the non-file parameters side of that tab.

File: nbextensions/bulkImportCell/tabs/configure.js [hunk 29]
```diff
@@ -176,7 +176,13 @@ define([
             fileTypePanel.updateState(getFileTypeState());
             model.setItem('state.selectedFileType', fileType);
             // stop and start the widgets.
-            return stopInputWidgets().then(startInputWidgets());
+            return stopInputWidgets()
+                .then(() => startInputWidgets())
+                .then(() => {
+                    cellBus.emit('toggled-active-filetype', {
```
Comment (briehl) [inline nbextensions/bulkImportCell/tabs/configure.js:182]: Added this to alert whatever's listening (mainly the test suite right now) that changes have been made and should be checked.

**Review by ialarmedalien — APPROVED** [review]
LGTM!
